### PR TITLE
Sync/Single cleanup

### PR DESCRIPTION
**Pull request by lydia-duncan**
(with help from @bradcray)

Previously to this effort, we leaked sync and single variables instead of
cleaning them up.  This has always been deemed undesirable, but I
was motivated to look at it now by a failing test on our cygwin test
box which was complaining about being unable to initialize a sync
variable due to a lack of resources.

Specifically, this effort redefines fifo's sync clean up function to call
pthread_cond_destroy and pthread_mutex_destroy.  It also removes
two places in the compiler (once in function resolution and once in
call destructors) where auto destroys were removed if the variable
in question was a sync variable, and removes a minor hack in
call destructors which reinserts destruction for syncs defined in our
internal and standard modules.  Because this hack was no longer
necessary, our reduction destructor needed to be defined to delete
its sync field.

The impact is that because syncs will be cleaned up when they
leave scope, the correct way to use sync variables in a parallel
context is to define them outside of a sync statement which
encapsulates their use.  Otherwise, a task spawned to use them
is not guaranteed to run before their destructor is called.  Several
tests have been re-written to use this model instead.

This also means that the correct way to return a sync variable from
a function is undefined.  The test of this functionality is now a
semantic future until we decide on expected behavior.

Tested over full qthreads and fifo, and valgrind testing with
release/, studies/, performance/, reductions/, types/single, and
types/sync/

### CODE REVIEW

**Comment by lydia-duncan** [comment]
@hildeth is probably the best person to review this, when he gets the chance.

**Comment by hildeth** [comment]
I opened this and reviewing it looks like real work. I'll give it a thorough review when I return on Wednesday.


Sent on a Samsung Galaxy S®4


-------- Original message --------
From: Lydia Duncan
Date:05/14/2015 6:02 PM (GMT-05:00)
To: chapel-lang/chapel
Cc: Tom Hildebrandt
Subject: Re: [chapel] Sync/Single cleanup (#1955)


@hildeth<https://github.com/hildeth> is probably the best person to review this, when he gets the chance.

File: test/parallel/sync/diten/returnSync.future [hunk 1]
```diff
@@ -0,0 +1,8 @@
+semantic: Should we be able to return a sync variable?
+
+The change to destroy sync variables when they leave scope makes returning
+syncs problematic, especially since copying syncs is not recommended.  Should
+it be possible to return a sync which would otherwise be cleaned up?
+
+When resolving this future, please remove the skipif and test with
+TASKS=qthreads
```
Comment (hildeth) [inline test/parallel/sync/diten/returnSync.future]: Missing EOL.

File: test/parallel/sync/figueroa/ReadMethods.chpl [hunk 2]
```diff
@@ -8,24 +8,27 @@ proc foo(type t, u: t, v: t, name) {
...
+      done = true;
+      write  ("2: value has changed to ", s.readFF());
+      writeln(" and it is ", if s.isFull then "full" else "empty");
+      sleep(1);
+      writeln("2: after sleeping, value is still ", s.readXX());
+      s.reset();
+      writeln("2: value has been reset to ", s.readFE());
+      done = true;
+    }
     sleep(1);
```
Comment (hildeth) [inline test/parallel/sync/figueroa/ReadMethods.chpl]: Not that you wrote this, but a one-second delay is a glacial age in computer time.  Using
  usleep(100000) instead would let the program run faster.

**Comment by hildeth** [comment]
Overall: Yippee!!!
I think we've been waiting for this regularization of the behavior of syncs/singles for a long time.
Okay to commit as-is or with minor edits as suggested.

File: test/parallel/sync/diten/returnSync.future (hunk 1 above)
Comment (lydia-duncan) [inline test/parallel/sync/diten/returnSync.future]: I ALWAYS do that!  You'd think I'd remember by now.

**Comment by lydia-duncan** [comment]
@bradcray, did you want to look at it before the merge?

**Comment by bradcray** [comment]
Nope, go for it.

**Pull request merged**